README.md
# sslexpired: an HTTP(S) service to check if an SSL certificate will expire soon
[![Build Status](https://travis-ci.org/shaftoe/sslexpired.svg?branch=master)](https://travis-ci.org/shaftoe/sslexpired)
[![License](https://img.shields.io/badge/license-Apache--2.0-blue.svg)](http://www.apache.org/licenses/LICENSE-2.0)
[![Go Report Card](https://goreportcard.com/badge/github.com/shaftoe/sslexpired)](https://goreportcard.com/report/github.com/shaftoe/sslexpired)
[![Issue Count](https://codeclimate.com/github/shaftoe/sslexpired/badges/issue_count.svg)](https://codeclimate.com/github/shaftoe/sslexpired)

[_sslexpired_][5] is a small [serverless][4] project written in [Go][1] for the [OpenWhisk][2] platform to quickly verify if an SSL certificate hosted at `https://<your_domain_here>` is going to expire any soon.

## Usage

By default, response will contain an `alert:true` key if the certificate is expiring in less then 30 days (or if the host is not mentioned between the valid ones):

    $ curl https://sslexpired.info/google.com
    {
      "notAfter": "2017-05-17 08:58:00 +0000 UTC",
      "daysTolerance": 30,
      "host": "google.com",
      "response": "SSL certificate for google.com will expire in 72 days",
      "validHosts": ["*.android.com", [...cut...],
      "daysLeft": 72
    }

The `days` parameter can be used to tweak the check tolerance, for example this request will alert because the certificate is expiring in 72 days and we ask for at least 100 days of validity:

    $ curl https://sslexpired.info/google.com?days=100
    {
      "notAfter": "2017-05-17 08:58:00 +0000 UTC",
      "daysTolerance": 100,
      "alert": true,
      "host": "google.com",
      "response": "SSL certificate for google.com will expire in 72 days",
      "validHosts": ["*.android.com", [...cut...],
      "daysLeft": 72
    }

## Current limitations

- checks for expiration timestamp only, doesn't actual verify SSL validity
- only DNS names supported (i.e. no IP addresses)

## Develop on OpenWhisk

Fetch the development environment installing [OpenWhisk development Vagrant box][3] and set up credentials for `wsk` tool as suggested

### Deploy sslexpired action and api gateway

    $ ./build.sh create
    # ... edit code ...
    $ ./build.sh update

### Invoke sslexpired action

    $ ./build.sh run <project_name>

or fetch the api URL and use an http client

    $ wsk -i api-experimental list /sslexpired
    $ curl <https url>/?project=<project_name>

### Destroy sslexpired

    $ ./build.sh delete

[1]: https://golang.org/ "Go"
[2]: http://openwhisk.org/ "OpenWhisk"
[3]: https://github.com/openwhisk/openwhisk#quick-start "OpenWhisk devel quick start"
[4]: https://en.wikipedia.org/wiki/Serverless_computing
[5]: https://sslexpired.info/


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.1.1] - 2017-03-06
### Added
- issuedBy key (and tests)

## [0.1.0] - 2017-03-04
### Added
- first draft of the Go app
- README
- LICENCE
- build script (build.sh)


sslexpired_test.go
package main

import (
	"bytes"
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"crypto/x509/pkix"
	"fmt"
	"io/ioutil"
	"math/big"
	"reflect"
	"regexp"
	"sort"
	"testing"
	"time"
)

func TestChangelogUpToDate(t *testing.T) {
	filename := "CHANGELOG.md"
	match := []byte("## [" + version + "]")
	sr := `## \[(\d+\.)*(\d+)\]`

	changelog, err := ioutil.ReadFile(filename)
	if err != nil {
		t.Error("Can not read", filename)
	}

	curIndex := bytes.Index(changelog, match)
	if curIndex < 0 {
		t.Error("Entry for version", version, "not found in", filename)
	}

	r, err := regexp.Compile(sr)
	if err != nil {
		t.Error("Can not compile regexp", sr)
	}

	firstIndex := r.FindIndex(changelog)
	switch {
	case firstIndex == nil:
		t.Error("Could not find any string matching", sr, "in", filename)
	case curIndex > firstIndex[0]:
		t.Error(version, "is not the latest entry in", filename)
	}
}

// mockCert creates an SSL certificate with expiration timestamp 30 days in the future
func mockCert(host string, altDNSNames []string, validForDays int) (*x509.Certificate, error) {
	daysInHours := time.Duration(24 * validForDays)
	notBefore := time.Now()
	notAfter := notBefore.Add(daysInHours * time.Hour)
	priv, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return nil, err
	}
	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
	if err != nil {
		return nil, err
	}
	template := &x509.Certificate{
		Subject:      pkix.Name{CommonName: host},
		DNSNames:     altDNSNames,
		SerialNumber: serialNumber,
		NotBefore:    notBefore,
		NotAfter:     notAfter,
	}
	derBytes, err := x509.CreateCertificate(rand.Reader, template, template, &priv.PublicKey, priv)
	if err != nil {
		return nil, err
	}
	cert, err := x509.ParseCertificate(derBytes)
	if err != nil {
		return nil, err
	}
	return cert, nil
}

func TestValidateInput(t *testing.T) {
	var msg map[string]interface{}
	var ok bool
	var args []string
	var text, expected string

	args = []string{"cmdName"}
	msg, ok = validateInput(args, 30)
	expected = "wrong number of arguments: expected 1 received 0"
	if ok || msg["err"] != expected {
		t.Error("Empty args should make validateInput fail with ->", expected)
	}

	args = []string{"cmdName", "", ""}
	msg, ok = validateInput(args, 0)
	expected = "wrong number of arguments: expected 1 received 2"
	if ok || msg["err"] != expected {
		t.Error("Empty args should make validateInput fail with ->", expected)
	}

	text = "not a json"
	args = []string{"cmdName", text}
	msg, ok = validateInput(args, 30)
	if ok || msg["err"] == nil {
		t.Error("validate fails if input is not json marshable")
	}

	text = `{"something":"wrong"}`
	args = []string{"cmdName", text}
	msg, ok = validateInput(args, 30)
	expected = "malformed request: missing host string"
	if ok || msg["err"] != expected {
		t.Error("host key is mandatory")
	}

	text = `{"days":"10"}`
	args = []string{"cmdName", text}
	msg, ok = validateInput(args, 10)
	expected = "malformed request: missing host string"
	if ok || msg["err"] != expected {
		t.Error("host key is mandatory")
	}

	text = `{"host":"ok","days":"notAnInteger"}`
	args = []string{"cmdName", text}
	_, ok = validateInput(args, 0)
	if ok {
		t.Error("should fail if days key string is not parsable to int")
	}

	text = `{"host":"ok","days":0}`
	args = []string{"cmdName", text}
	_, ok = validateInput(args, 123)
	if ok {
		t.Error("should fail if days key is not in string format")
	}

	text = `{"days":"1234","host":"someHost"}`
	args = []string{"cmdName", text}
	msg, ok = validateInput(args, 0)
	if !ok || msg["daysTolerance"] != 1234 || msg["host"] != "someHost" {
		t.Error("something wrong validating", text, "received", msg)
	}
}

func TestHostInCert(t *testing.T) {
	var domain, host, wild string
	var cert *x509.Certificate
	var validForDays = 30

	f := func(h string, c *x509.Certificate) {
		if !hostInCert(h, c) {
			t.Errorf("%s should be a valid host string", h)
		}
	}

	domain = "somethingstrange.com"
	host = "www." + domain
	wild = "*." + domain
	cert, _ = mockCert(host, []string{wild, domain}, validForDays)
	if hostInCert("bogus.com", cert) {
		t.Error("bogus domain should not be a valid host string")
	}
	for _, h := range []string{domain, host, wild} {
		f(h, cert)
	}

	domain = "why.not.some.more.levels.com."
	host = "www." + domain
	wild = "*." + domain
	cert, _ = mockCert(host, []string{wild, domain}, validForDays)
	for _, h := range []string{domain, host, wild} {
		f(h, cert)
	}
}

func TestParsedMsg(t *testing.T) {
	out := parsedMsg(map[string]interface{}{"msg": "notajson"})
	if out != `{"msg":"notajson"}` {
		t.Error("failed unmarshaling map")
	}
}

func TestValidateSSL(t *testing.T) {
	var msg, out map[string]interface{}
	var delta int
	var cert *x509.Certificate
	var expected, host, hostAlt string
	var validHosts []string

	// SSL cert created by mockCert is expiring in 30 days minus a few nanoseconds,
	// so the expected output form our API will be 29 days
	host = "some.host.here.com"
	hostAlt = "*.another.wildcard.domain"
	validHosts = []string{host, hostAlt, "another.unrelated.com", "whatever.else."}
	sort.Strings(validHosts)
	delta = 30
	cert, _ = mockCert(host, validHosts, delta)
	msg, _ = validateInput([]string{"command", `{"host":"` + host + `"}`}, delta)
	out = validateSSL(msg, cert)
	if out["alert"] == nil || out["alert"] != true {
		t.Error("check must fail when days left are smaller then delta, received", out)
	}
	if out["daysLeft"] == nil || out["daysLeft"] != delta-1 {
		t.Errorf("daysLeft must be %d, received %s", delta-1, out)
	}
	if out["host"] == nil || out["host"] != host {
		t.Error("host must be", host, "received", out)
	}
	if out["daysTolerance"] == nil || out["daysTolerance"] != delta {
		t.Error("daysTolerance must be default of ", delta, "received", out)
	}
	if out["notAfter"] == nil {
		t.Error("notAfter key must be present:", out)
	}
	expected = fmt.Sprintf("SSL certificate for %s will expire in %d days", host, 29)
	if out["response"] == nil || out["response"] != expected {
		t.Error("expected", expected, "have", out["response"])
	}
	if !reflect.DeepEqual(out["validHosts"], validHosts) {
		t.Errorf("#validHosts: want %s got %s", validHosts, out["validHosts"])
	}
	if out["issuedBy"] == nil || out["issuedBy"] != "" {
		t.Error("issuedBy not found in keys")
	}

	// this time we check the DNSAlt section, and we use a smaller delta, so
	// it should not alert
	host = "www.another.wildcard.domain"
	delta = 29
	msg, _ = validateInput([]string{"command", fmt.Sprintf(`{"host":"%s","days":"%d"}`, host, delta)}, 0)
	out = validateSSL(msg, cert)
	if out["alert"] != nil {
		t.Error("check must succeed when days left are bigger then delta,", out)
	}
	if out["daysLeft"] == nil || out["daysLeft"] != delta {
		t.Error("daysLeft must be 29, received", out)
	}
	if out["host"] == nil || out["host"] != host {
		t.Error("host must be", host, "received", out)
	}
	if out["daysTolerance"] == nil || out["daysTolerance"] != delta {
		t.Error("daysTolerance must be", delta, "received", out)
	}
	if out["notAfter"] == nil || out["response"] == nil {
		t.Error("both response and notAfter keys must be present:", out)
	}
	expected = fmt.Sprintf("SSL certificate for %s will expire in %d days", host, 29)
	if out["response"] == nil || out["response"] != expected {
		t.Error("expected", expected, "have", out["response"])
	}
	if !reflect.DeepEqual(out["validHosts"], validHosts) {
		t.Errorf("#validHosts: want %s got %s", validHosts, out["validHosts"])
	}

	host = "a.bogus.domain"
	msg, _ = validateInput([]string{"command", fmt.Sprintf(`{"host":"%s"}`, host)}, 0)
	out = validateSSL(msg, cert)
	expected = fmt.Sprintf("host %s not valid for the SSL certificate", host)
	if out["alert"] == nil || out["err"] == nil || out["err"] != expected {
		t.Error("validate should fail for", host, "received", out)
	}
}


build.sh
#!/bin/bash
# A few wrappers around the OpenWhisk platform.

set -e
readonly action='sslexpired'
readonly command="${1:-create}"
readonly api="/${action}"
readonly version="v1"
readonly wskCli="wsk -i" # remove -i if developing on Bluemix platform

cleanup() {
    rm -f action.zip exec
}
trap cleanup ERR EXIT

usage() {
    echo "Usage: $0 [create|update|run|delete]"
}

build() {
    rm -f action.zip exec
    env GOOS=linux GOARCH=amd64 go build -o exec "${action}.go"
    zip action.zip exec
}

createupdate() {
    $wskCli action $command $action action.zip --docker
}

run() {
    $wskCli action invoke $action --blocking --result --param host $2 --param days $3
}

createapigateway() {
    $wskCli api-experimental create $api /$version get $action
}

delete() {
    $wskCli api-experimental delete $api
    $wskCli action delete $action
}

case $1 in
    create) set -x; build; createupdate; createapigateway;;
    update) set -x; build; createupdate;;
    run) set -x; run $*;;
    delete) set -x; delete;;
    *) usage; exit 1;;
esac


sslexpired.go
package main

import (
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

const version = "0.1.1"

// SSLCerts returns a slice of x509 Certificates available at the
// target hostname
func SSLCerts(host string) ([]*x509.Certificate, error) {
	conf := &tls.Config{
		InsecureSkipVerify: true,
	}
	dialer := &net.Dialer{
		Timeout: 5 * time.Second,
	}
	conn, err := tls.DialWithDialer(dialer, "tcp", host+":443", conf)
	if err != nil {
		return nil, err
	}
	conn.Close()
	return conn.ConnectionState().PeerCertificates, nil
}

func validateInput(args []string, delta int) (map[string]interface{}, bool) {
	msg := make(map[string]interface{})

	// native OpenWhisk actions receive one argument, the JSON object as a string
	if len(args) != 2 {
		msg["err"] = fmt.Sprintf("wrong number of arguments: "+
			"expected 1 received %d", (len(args) - 1))
		return msg, false
	}

	// unmarshal the string to a JSON object
	var obj map[string]interface{}
	err := json.Unmarshal([]byte(args[1]), &obj)
	if err != nil {
		msg["err"] = err.Error()
		return msg, false
	}

	// host parameter IS mandatory
	host, ok := obj["host"].(string)
	if !ok || host == "" {
		msg["err"] = "malformed request: missing host string"
		return msg, false
	}
	msg["host"] = host

	// days parameter is not mandatory, we use default if not present
	days, ok := obj["days"]
	if ok {
		switch days.(type) {
		case string:
			d, err := strconv.Atoi(days.(string))
			if err != nil {
				msg["err"] = "failed to parse days parameter: " + err.Error()
				return msg, false
			}
			msg["daysTolerance"] = d
		default:
			msg["err"] = fmt.Sprintf("days parameter malformatted: %T, must be string", days)
			return msg, false
		}
	} else {
		msg["daysTolerance"] = delta
	}

	return msg, true
}

func daysLeft(cert *x509.Certificate) float64 {
	return cert.NotAfter.Sub(time.Now()).Hours() / 24
}

func issuedBy(cert *x509.Certificate) string {
	if len(cert.Issuer.Organization) > 0 {
		return cert.Issuer.Organization[0]
	}
	return ""
}

func matchFound(host string, cert *x509.Certificate) bool {
	if cert.Subject.CommonName == host {
		return true
	}
	for _, name := range cert.DNSNames {
		if host == name {
			return true
		}
	}
	return false
}

func allCertHosts(cert *x509.Certificate) []string {
	set := map[string]struct{}{}
	set[cert.Subject.CommonName] = struct{}{}
	for _, name := range cert.DNSNames {
		set[name] = struct{}{}
	}

	var hosts []string
	for name := range set {
		hosts = append(hosts, name)
	}
	sort.Strings(hosts)

	return hosts
}

func hostInCert(host string, cert *x509.Certificate) bool {
	if matchFound(host, cert) {
		return true
	}

	// if exact match for host is not found, we search for a wildcard
	// replacing the first part of the hostname with *
	splitted := strings.Split(host, ".")
	if len(splitted) > 2 {
		splitted[0] = "*"
		return matchFound(strings.Join(splitted, "."), cert)
	}

	return false
}

func parsedMsg(msg map[string]interface{}) string {
	resp, err := json.Marshal(msg)
	if err != nil {
		return fmt.Sprintf(`{"err":"%s"}`, err)
	}
	return fmt.Sprint(string(resp))
}

func validateSSL(msg map[string]interface{}, cert *x509.Certificate) map[string]interface{} {
	dLeft := daysLeft(cert)
	msg["daysLeft"] = int(dLeft) // we round the number to biggest integer
	msg["response"] = fmt.Sprintf("SSL certificate for %s will expire in %d days",
		msg["host"], msg["daysLeft"])
	msg["notAfter"] = cert.NotAfter.String()
	msg["validHosts"] = allCertHosts(cert)
	msg["issuedBy"] = issuedBy(cert)

	if dLeft < float64(msg["daysTolerance"].(int)) {
		msg["alert"] = true
	}

	if !hostInCert(msg["host"].(string), cert) {
		msg["alert"] = true
		msg["err"] = fmt.Sprintf("host %s not valid for the SSL certificate", msg["host"])
	}

	return msg
}

func main() {
	msg, ok := validateInput(os.Args, 30)
	if !ok {
		fmt.Println(parsedMsg(msg))
		return
	}
	certs, err := SSLCerts(msg["host"].(string))
	if err != nil {
		delete(msg, "daysTolerance")
		msg["err"] = err.Error()
		fmt.Println(parsedMsg(msg))
		return
	}
	if len(certs) == 0 {
		msg["err"] = "certificates list is empty"
		fmt.Println(parsedMsg(msg))
		return
	}
	// We check only the first certificate in the chain
	fmt.Println(parsedMsg(validateSSL(msg, certs[0])))
}
